subprojects/internal-testing/src/main/groovy/org/gradle/util/internal/Resources.java

```diff
@@ -45,22 +61,89 @@ public TestFile getResource(String name) {
 
     /**
      * Locates the resource with the given name, relative to the current test class.
-     * @return the resource, or null if not found.
+     *
+     * Can also handle extracting resources contained within a jar file found in the same manner.
+     *
+     * @return the resource, or {@code null} if not found
      */
     public TestFile findResource(String name) {
         assertNotNull(testClass);
         URL resource = testClass.getResource(name);
         if (resource == null) {
             return null;
         }
-        assertEquals(String.format("Cannot handle resource URI %s", resource), "file", resource.getProtocol());
-        File file;
+
+        switch (resource.getProtocol()) {
+            case "jar":
+                return fromWithinJar(resource);
+            case "file":
+                return fromFile(resource);
+            default:
+                throw new RuntimeException(String.format("Cannot handle resource URI %s", resource));
+        }
+    }
+
+    @Nonnull
+    private TestFile fromFile(URL resource) {
         try {
-            file = new File(resource.toURI());
+            return new TestFile(resource.toURI());
         } catch (URISyntaxException e) {
             throw new RuntimeException(e);
         }
-        return new TestFile(file);
+    }
+
+    @Nonnull
+    private TestFile fromWithinJar(URL resourceUrl) {
+        try {
+            int indexOfJarSeparator = resourceUrl.getPath().indexOf("!/");
+            String jarFilePath = resourceUrl.getPath().substring(5, indexOfJarSeparator);
+            File outputDir;
+
+            synchronized (extractedJars) {
+                if (!extractedJars.containsKey(testClass)) {
+                    extractedJars.put(testClass, new HashMap<>());
+                }
+
+                Map<String, File> extractedJarsForTestClass = extractedJars.get(testClass);
+                if (!extractedJarsForTestClass.containsKey(jarFilePath)) {
+                    outputDir = Files.createTempDir();
```

Comment:
I think we should try to find a way to extract this into the final destination vs copying it to the system level tmp directory first.